core/src/toga/documents.py

```diff
@@ -118,5 +169,313 @@ def create(self) -> None:
+            path = Path(path_or_index).resolve()
+        for item in self.elements:
+            if item.path == path:
+                return item
+
+        # No match found
+        raise KeyError(path_or_index)
+
+    def _add(self, document: Path):
+        if document in self:
+            raise ValueError("Document is already being managed.")
+
+        self.elements.append(document)
+
+    def _remove(self, document: Path):
+        if document not in self:
+            raise ValueError("Document is not being managed.")
+
+        self.elements.remove(document)
+
+    def new(self, document_type: type[Document]) -> Document:
+        """Create a new document of the given type, and show the document window.
+
+        :param document_type: The document type that has been requested.
+        :returns: The newly created document.
+        """
+        document = document_type(app=self.app)
+        document.show()
+        return document
+
+    async def request_open(self) -> Document:
+        """Present a dialog asking the user for a document to open, and show the window
+        for that document.
+
+        If the provided path is already an open document, the existing representation
+        for the document will be given focus.
+
+        :returns: The document that was opened.
+        :raises ValueError: If the path describes a file that is of a type that doesn't
+            match a registered document type.
+        """
+        # A safety catch: if app modal dialogs aren't actually modal (eg, macOS) prevent
+        # a second open dialog from being opened when one is already active. Attach the
+        # dialog instance as a private attribute; delete as soon as the future is
+        # complete.
+        if hasattr(self, "_open_dialog"):
+            return
+
+        # CLOSE_ON_LAST_WINDOW is a proxy for the GTK/Windows behavior of loading content
+        # into the existing window. This is actually implemented by creating a new window
+        # and disposing of the old one; mark the current window for cleanup
+        if self.app._impl.CLOSE_ON_LAST_WINDOW:
+            if hasattr(self.app.current_window, "_commit"):
+                self.app.current_window._replace = (
+                    await self.app.current_window._commit()
+                )
+
+        self._open_dialog = dialogs.OpenFileDialog(
+            self.app.formal_name,
+            file_types=(
+                list(self.app.document_types.keys())
+                if self.app.document_types
+                else None
+            ),
+        )
+        path = await self.app.dialog(self._open_dialog)
+        del self._open_dialog
+
+        if path:
+            return self.open(path)
+        else:
+            # If the open was cancelled, but the current window has a replacement
+            # marker, remove the replacement marker
+            if hasattr(self.app.current_window, "_replace"):
+                del self.app.current_window._replace
```

Comment:
It would be clearer if this marker was cleaned up in one and only one place, in the same method that created it. This could be done in a `finally` block, with the help of a variable saving the original value of `current_window`. The similar block at the bottom of `open` can then be removed.